File: extension/src/score/textEmbedding.ts
import { fnv1a64 } from "./textNearDuplication";


export const EMBEDDING_DIMENSIONS = 256;

const MASK_64 = (1n << 64n) - 1n;
const SIGN_BIT = 63n;

const ALPHANUMERIC = /[\p{L}\p{N}]/u;

export function tokenize(text: string): string[] {
  const tokens: string[] = [];
  let current = "";
  for (const character of text.toLowerCase()) {
    if (ALPHANUMERIC.test(character)) {
      current += character;
      continue;
    }
    if (current.length > 0) {
      tokens.push(current);
      current = "";
    }
  }
  if (current.length > 0) {
    tokens.push(current);
  }
  return tokens;
}

export function terms(tokens: readonly string[]): string[] {
  const result = [...tokens];
  for (let i = 0; i + 1 < tokens.length; i++) {
    result.push(`${tokens[i]} ${tokens[i + 1]}`);
  }
  return result;
}

export type TermCounts = number[];

export function hashTerms(text: string, dimensions = EMBEDDING_DIMENSIONS): TermCounts {
  const counts = new Map<number, number>();
  for (const term of terms(tokenize(text))) {
    const hash = fnv1a64(term) & MASK_64;
    const bucket = Number(hash % BigInt(dimensions));
    const sign = (hash >> SIGN_BIT) & 1n ? -1 : 1;
    counts.set(bucket, (counts.get(bucket) ?? 0) + sign);
  }
  const flat: TermCounts = [];
  for (const bucket of [...counts.keys()].sort((a, b) => a - b)) {
    flat.push(bucket, counts.get(bucket) as number);
  }
  return flat;
}

export function embedTermCounts(
  counts: TermCounts,
  dimensions = EMBEDDING_DIMENSIONS,
): number[] | null {
  const vector = new Array<number>(dimensions).fill(0);
  for (let i = 0; i + 1 < counts.length; i += 2) {
    const bucket = counts[i] as number;
    if (bucket < 0 || bucket >= dimensions) {
      return null;
    }
    vector[bucket] = counts[i + 1] as number;
  }
  return normalizeVector(vector);
}

export function embedText(text: string, dimensions = EMBEDDING_DIMENSIONS): number[] | null {
  return embedTermCounts(hashTerms(text, dimensions), dimensions);
}

export function cosineSimilarity(a: readonly number[], b: readonly number[]): number {
  let total = 0;
  for (let i = 0; i < a.length; i++) {
    total += (a[i] as number) * (b[i] as number);
  }
  return total;
}

export function normalizeVector(vector: readonly number[]): number[] | null {
  let sumOfSquares = 0;
  for (const value of vector) {
    sumOfSquares += value * value;
  }
  if (sumOfSquares === 0) {
    return null;
  }
  const norm = Math.sqrt(sumOfSquares);
  return vector.map((value) => value / norm);
}

File: research/verdict_research/features/text_embedding.py
import math

from verdict_research.features.text_near_duplication import MASK_64, fnv1a64

EMBEDDING_DIMENSIONS = 256
_SIGN_BIT = 63


def tokenize(text: str) -> list[str]:
    tokens = []
    current = ""
    for character in text.lower():
        if character.isalnum():
            current += character
            continue
        if current:
            tokens.append(current)
            current = ""
    if current:
        tokens.append(current)
    return tokens


def terms(tokens: list[str]) -> list[str]:
    result = list(tokens)
    result.extend(f"{tokens[i]} {tokens[i + 1]}" for i in range(len(tokens) - 1))
    return result


def hash_terms(text: str, dimensions: int = EMBEDDING_DIMENSIONS) -> list[int]:
    counts: dict[int, int] = {}
    for term in terms(tokenize(text)):
        hash_value = fnv1a64(term) & MASK_64
        bucket = hash_value % dimensions
        sign = -1 if (hash_value >> _SIGN_BIT) & 1 else 1
        counts[bucket] = counts.get(bucket, 0) + sign
    flat: list[int] = []
    for bucket in sorted(counts):
        flat.extend((bucket, counts[bucket]))
    return flat


def normalize_vector(vector: list[float]) -> list[float] | None:
    sum_of_squares = 0.0
    for value in vector:
        sum_of_squares += value * value
    if sum_of_squares == 0:
        return None
    norm = math.sqrt(sum_of_squares)
    return [value / norm for value in vector]


def embed_term_counts(
    counts: list[int], dimensions: int = EMBEDDING_DIMENSIONS
) -> list[float] | None:
    vector = [0.0] * dimensions
    for i in range(0, len(counts) - 1, 2):
        bucket = counts[i]
        if bucket < 0 or bucket >= dimensions:
            return None
        vector[bucket] = counts[i + 1]
    return normalize_vector(vector)


def embed_text(text: str, dimensions: int = EMBEDDING_DIMENSIONS) -> list[float] | None:
    return embed_term_counts(hash_terms(text, dimensions), dimensions)


def cosine_similarity(a: list[float], b: list[float]) -> float:
    total = 0.0
    for i in range(len(a)):
        total += a[i] * b[i]
    return total
